### ISSUE TITLE
Licensing

### ISSUE THREAD

**Issue by muraiki** [alkaline-ml/pmdarima#419]
With the disclaimer that I'm not an open source software license expert, I noticed that this project is MIT licensed but has sections based on GPLv2 code from the R library `tseries`. I think this could make this project a derived work, meaning that it would need to follow the terms of GPLv2.

In particular, in issue #67 @tgsmith61591 links to the `tseries::adf.test` [source code](https://github.com/cran/tseries/blob/master/R/test.R#L141) in explaining how `ADFTest` was written, and then further explains how code from the `tseries` package was used to inform the development of some functions in `pmdarima`.

The issue of determining whether something is a derived work is complicated, with a good summary of what's been determined in various courts being present in [this article](https://copyleft.org/guide/comprehensive-gpl-guidech5.html). It's not my intention to make this determination myself, but since I need to consider the impact of software licenses when I use libraries, I thought that it'd be prudent to create an issue for this and hear what the author and others have to say. I apologize in advance if I've misunderstood anything.

**Comment by tgsmith61591** [comment]
Hey @muraiki thanks for the question. I am not really sure writing code in a different language that is loosely based on an implementation in a different language would be considered a derivative work. However, I'm not a lawyer. Is our license causing issues for you?

**Comment by muraiki** [comment]
It turns out that I should have checked the [GPL FAQ](https://www.gnu.org/licenses/gpl-faq.html#TranslateCode):

> What does the GPL say about translating some code to a different programming language? (#TranslateCode)

> Under copyright law, translation of a work is considered a kind of modification. Therefore, what the GPL says about modified versions applies also to translated versions. The translation is covered by the copyright on the original program.
> If the original program carries a free license, that license gives permission to translate it. How you can use and license the translated program is determined by that license. If the original program is licensed under certain versions of the GNU GPL, the translated program must be covered by the same versions of the GNU GPL.

As to whether the way you based your code on the R code constitutes creation of a derivative work, it sounds like that would be determined using the _Abstraction, Filtration, Comparison Test_ described in the copyleft [article](https://copyleft.org/guide/comprehensive-gpl-guidech5.html).

If this project does fall under the GPL, that would make using it at my workplace more difficult. There's a whole other murky legal area around whether using GPL'd libraries in an interpreter, such as R or Python, constitutes creation of a derivative work. See this [R-Bloggers post](https://www.r-bloggers.com/2019/01/how-gpl-makes-me-leave-r-for-python/) about a person whose employer apparently consulted with four different software lawyers. In the end, interpreting these applications of the GPL is difficult enough that I'd rather not expose my employer to them all.

**Comment by muraiki** [comment]
Also, I found a useful explanation of a similar situation in this [discussion](https://lobste.rs/s/q94fty/what_does_gpl_say_about_translating_some). The maintainer of GNU Octave (username JordiGH) wanted to translate GPL'd R code into Python to add to the statsmodels library.

**Comment by tgsmith61591** [comment]
I understand the risk aversion of large enterprises. Please let me know what license your workplace requires, and I can add it a la [darknet](https://github.com/pjreddie/darknet) 😄

**Comment by muraiki** [comment]
Sorry, I think I've been unclear! The current license is totally fine for usage in my workplace. What I'm concerned about is whether the current license can legally be applied, if pmdarima is actually a GPL-derived work by virtue of some portions of it being based on GPLv2 R code.

This has been on my mind because a [relicensing situation](https://github.com/rails/rails/issues/41750) recently affected Rails: Rails has a dependency on a library that had a non-GPL license applied, despite it being a GPL-derived work. This means that despite the library previously having been distributed under a non-GPL license, it actually was GPL in practice. The author did the legally required thing by [applying the GPL](https://github.com/mimemagicrb/mimemagic/issues/97) to all affected versions, which necessitated that Rails replace the dependency.

To summarize for the sake of clarity: if pmdarima is a GPL-derived work (perhaps through the Abstraction, Filtration, Comparison Test, which I'm not qualified to perform), then from the commit with R-based code onward, the current license is legally invalid and the GPLv2 must instead be applied. As for releasing a new version of pmdarima without the R-based code, I don't know how that would work in terms of reapplying the current non-GPL license.

Sorry if this is really confusing or if I'm making a mountain out of a molehill! After thinking about this, I think this might ultimately boil down to "a lawyer needs to figure it out", as it hinges upon whether the way you translated the R code (even "loosely") resulted in a derived work or not.

**Comment by tgsmith61591** [comment]
I'm afraid no one on the pmdarima team is qualified to make this judgment, it would seem. While portions of the code have been inspired by the R package, and it's designed to behave as closely to the R package as possible, it's fundamentally different in many ways and would up to a lawyer to determine whether or not it's GPL-derived.

I'm sorry if this does not really address your issue.

**Comment by muraiki** [comment]
There's no need to apologize. Thank you for taking the time to look into this! I really appreciate your team's work on making such a great library... and on making using Python as a former R user much easier!
